### PR TITLE
Enable BinaryFormatter killbit by default in aspnet applications; introduce feature switch

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Contributes to https://github.com/dotnet/sdk/issues/12217

This is the "feature switch" companion PR to https://github.com/dotnet/runtime/pull/38963. It also triggers the `BinaryFormatter` killbit by default for aspnet applications. Unlike wasm applications, the project file can still opt-in to allowing `BinaryFormatter` payloads to run.

Open questions and notes for reviewers:

- It'd be good to get clarification on the open question at https://github.com/dotnet/sdk/issues/12217#issuecomment-657151788.

- Is using the .targets file the right way to perform the check "the developer must have defined this element, and the value must be equal to _true_"? What I'm trying to avoid is having somebody write `<Enable>I like puppies!</Enable>" and having that re-enable the feature. In aspnet project types, the default behavior should be _killbit on_ unless somebody explicitly specifies _killbit off_.

/cc @davidfowl 

__This is a breaking change and should be documented in the Preview 8 release notes.__

This change should only affect .NET 5.0+ applications. Setting the switch has no effect in .NET Framework or .NET Core 2.x - 3.x.

Further documentation will be available at https://aka.ms/binaryformatter.

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
Note to reviewers: I have run this breaking change by both the runtime + the aspnet architects. There is signoff for Preview 8. Subject to PR approval, of course. :)

**Review by vijayrkn — APPROVED** [review]

**Comment by davidfowl** [comment]
Why is it a build property? We don't plan to do this for anything else right?

**Comment by GrabYourPitchforks** [comment]
All of the feature switches are csproj properties as far as I can tell. See the lines immediately above the ones I introduced in the Microsoft.Net.Sdk.targets file.

File: src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets [hunk 1]
```diff
@@ -35,6 +35,17 @@ Copyright (c) .NET Foundation. All rights reserved.
...
 
+  <PropertyGroup>
+    <!--
+      Unless the project explicitly set the property value to 'true' (case-insensitive),
+      default it to 'false'. Not specifying a value at all (or specifying an empty value)
+      will also default to 'false'. In .NET 5.0 this functionality is explicitly disabled
+      by default for web project types, and in future .NET releases the "disable by default"
+      logic will roll out to other project types.
+    -->
+    <EnableUnsafeBinaryFormatterSerialization Condition="'$(EnableUnsafeBinaryFormatterSerialization)' != 'true'">false</EnableUnsafeBinaryFormatterSerialization>
```
Comment (eerhardt) [inline src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:46]: How easily does this break an existing application? For example, if I put images in a `.resx` file is my app broken now?

Comment (GrabYourPitchforks) [reply src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets:46]: Correct, such an app will be broken. However:

* The overwhelming majority of web applications don't do such a thing; and
* The exception message text links to a document with clear remediation steps.

**Comment by eerhardt** [comment]
> It'd be good to get clarification on the open question at #12217 (comment).

I responded there, but copied the response here:

>> @eerhardt What's the expected behavior if a setting is specified in both the .csproj and the runtimeconfig.json file? 

> See the comment here:

https://github.com/dotnet/sdk/blob/a5f5bb5183517cb301050288b25c9c727e52e8a6/src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs#L163-L166

> So yes, the intention is that the settings in the .csproj will always win over a conflicting `runtimeconfig.template.json` 

_____

> Is using the .targets file the right way to perform the check "the developer must have defined this element, and the value must be equal to true"?

The general/typical MSBuild flow is:

* `.props` files get imported before a user's project.
* The user's project gets imported.
* `.targets` files get imported after a user's project.

Typically when you want a default for a property, you just check if it hasn't been set yet:

```xml
<PropertyGroup>
  <EnableUnsafeBinaryFormatterSerialization Condition="'$(EnableUnsafeBinaryFormatterSerialization)' == ''">false</EnableUnsafeBinaryFormatterSerialization>
</PropertyGroup>
```

In this specific case, the value of the MSBuild property is getting flown into the `runtimeconfig.json` file straight from the property. So in your example of `I like puppies`, you have to inspect what the runtime does when it sees an AppContext switch with the value `I like puppies`.

https://github.com/dotnet/runtime/blob/7b2364312f2145ffa3bad6e153459b412e1bf264/src/libraries/System.Private.CoreLib/src/System/AppContext.cs#L101

Says if the value doesn't parse into a `bool`, then it is treated the same as if the value wasn't there at all.

Note that this MSBuild logic isn't bullet-proof. All it would take is some other `.targets` file getting imported after this `.targets` file (or the user's project can get code executed after all `.targets` files if they really want) and overwriting the property to `I like puppies`. And now you are back to what you are trying to avoid.

My recommendation would be to not try to be fancy here. I would just default the value as normal. The safest place to put a check like that would be in the runtime itself.

**Review by eerhardt — APPROVED** [review]
Looks good.

You can decide if you want to keep the defaulting logic as is. It doesn't matter much.

**Comment by GrabYourPitchforks** [comment]
Yeah, this check wasn't intending to be overly fancy. It was trying to mimic what the long-term behavior would be across all project types, which is that the feature would default to disabled unless the literal string _true_ is present in the .csproj. The weirdness here comes from the fact that the core runtime still defaults the behavior to "enabled" unless the literal string _false_ is present in the .csproj.

Maybe best to back this out and say it's a problem for future me? :)

**Comment by eerhardt** [comment]
> Maybe best to back this out and say it's a problem for future me? :)

Honestly, I'm fine with either way. I guess if it were me I would have just written 

```xml
<PropertyGroup>
  <EnableUnsafeBinaryFormatterSerialization Condition="'$(EnableUnsafeBinaryFormatterSerialization)' == ''">false</EnableUnsafeBinaryFormatterSerialization>
</PropertyGroup>
```

> The weirdness here comes from the fact that the core runtime still defaults the behavior to "enabled" unless the literal string false is present in the .csproj.

Could change the 5.0 runtime to throw for this setting if the value doesn't parse to a bool?

**Comment by davidfowl** [comment]
> All of the feature switches are csproj properties as far as I can tell

No not even close. The important ones get hoisted. We have a million switches only a few of them are expressabls in msbuild (unless I missed something)

**Comment by eerhardt** [comment]
>> All of the feature switches are csproj properties as far as I can tell

> No not even close. The important ones get hoisted. We have a million switches only a few of them are expressabls in msbuild (unless I missed something)

See the [feature switch design doc](https://github.com/dotnet/designs/blob/master/accepted/2020/feature-switch.md) and [Libraries Feature Switches doc](https://github.com/dotnet/runtime/blob/master/docs/workflow/trimming/feature-switches.md) for reasonings why these are put into MSBuild.

> The important ones get hoisted

I think this one is important enough to get hoisted. How else would the ASP.NET SDK set it by default otherwise?

**Comment by GrabYourPitchforks** [comment]
@eerhardt 
I like your simpler _Condition_ proposal. Thanks for the advice. :)
Will get a new iteration out soon.

**Comment by davidfowl** [comment]
> I think this one is important enough to get hoisted. How else would the ASP.NET SDK set it by default otherwise?

It's not for ASP.NET Core, it's for user code. If we think people are using this and need to undo it then I buy it, is that the case here?

**Comment by eerhardt** [comment]
> If we think people are using this and need to undo it then I buy it, is that the case here?

There are multiple levels of why this is useful:

1. If I have an ASP.NET Core application that uses BinaryFormatter somewhere and I need to re-enable BinaryFormatter after upgrading to 5.0.
2. If I'm NOT in an ASP.NET Core application and I want to disable BinaryFormatter in my app, I can just set this property.
3. Sometime after 5.0, we are going to flip this property to be enabled by default for all app types. If a user has an app and they still want to use BinaryFormatter, they would set this property.

Do you have an alternative suggestion to what Levi has here?

**Comment by davidfowl** [comment]
> Do you have an alternative suggestion to what Levi has here?

No my beef is about the "importance" of a top level property like this to affect a single API (BinaryFormatter in this case). If this becomes our general strategy, we'll end up with *LOTS* of msbuild properties to affect runtime APIs. I know it's a slippery slope argument but I worry about things like this.

Maybe in a future we can entertain a code based option for enabling feature flags like this.

**Comment by GrabYourPitchforks** [comment]
Honestly I'd think most people would prefer to use _runtimeconfig.json_ for this. Using .csproj properties feels weird to me, but it is how the "feature switch" spec is defined.

**Comment by GrabYourPitchforks** [comment]
Apparently I fail very hard at merging. 🤦‍♂️

**Comment by eerhardt** [comment]
> Honestly I'd think most people would prefer to use runtimeconfig.json for this. Using .csproj properties feels weird to me, but it is how the "feature switch" spec is defined.

The issue is the setting needs to affect more than just the runtime. If you want the linker (or any other built-time process) to have knowledge of the setting, then it needs to be in MSBuild. Teaching the various build-time processes to run after the generation of runtimeconfig.json, and to know how to read/interpret the runtimeconfig.json just doesn't work.

**Pull request merged**